client/src/modules/auphonic/components/FileChooser.vue

```diff
@@ -0,0 +1,194 @@
+            <label for="audio_external_file" class="block text-sm font-medium text-gray-700"
+              >File</label
+            >
+            <select
+              v-model="fileSelection"
+              name="audio_external_file"
+              id="audio_external_file"
+              class="mt-1 block w-full pl-3 pr-10 py-2 text-base border-gray-300 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm rounded-md"
+              @input="handleExternalFileSelection"
+            >
+              <option v-for="file in serviceFiles" :key="file" :value="file">
+                {{ file }}
+              </option>
+            </select>
+          </div>
+          <div v-else>...</div>
+        </div>
+      </div>
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+// NEXT UP: returning picked values from FileChooser to parent component so it can be converted into API params
+import { ref, defineComponent, nextTick } from 'vue'
+import Module from '@components/module/Module.vue'
+import PodloveButton from '@components/button/Button.vue'
+import { PlusSmIcon } from '@heroicons/vue/outline'
+import { selectors } from '@store'
+
+import { injectStore, mapState } from 'redux-vuex'
+import * as auphonic from '@store/auphonic.store'
+import { Service } from '@store/auphonic.store'
+import SelectProduction from './components/SelectProduction.vue'
+import ManageProductionForm from './components/ManageProductionForm.vue'
+import AuphonicLogo from './components/Logo.vue'
+
+import {
+  Listbox,
+  ListboxButton,
+  ListboxLabel,
+  ListboxOption,
+  ListboxOptions,
+} from '@headlessui/vue'
+import { CheckIcon, SelectorIcon } from '@heroicons/vue/solid'
+
+export default defineComponent({
+  emits: ['update:modelValue'],
+
+  components: {
+    Listbox,
+    ListboxButton,
+    ListboxLabel,
+    ListboxOption,
+    ListboxOptions,
+    CheckIcon,
+    SelectorIcon,
+  },
+
+  data() {
+    return {
+      currentServiceSelection: null,
+      fileSelection: null,
+      urlValue: null,
+      fileValue: null,
+    }
+  },
+
+  setup() {
+    const state = mapState({
+      services: selectors.auphonic.incomingServices,
+      serviceFiles: selectors.auphonic.serviceFiles,
+    })
+
+    return {
+      state,
+      dispatch: injectStore().dispatch,
+    }
+  },
+
+  methods: {
+    async handleExternalFileSelection() {
+      await nextTick()
+      this.emitValue()
+    },
+    emitValue() {
+      const service = this.currentService
+      let value = null
+
+      switch (service.type) {
+        case 'url':
+          value = this.urlValue
+          break
+        case 'file':
+          value = this.fileValue
+          break
+
+        default:
+          value = this.fileSelection
+          break
+      }
+
+      this.$emit('update:modelValue', { service, value })
+    },
+  },
+
+  watch: {
+    ready() {
+      this.currentServiceSelection = this.state.services[0]?.uuid
```

Comment:
wenn du die current service selection in den store verlegst kannst du das alles in einer saga behandeln, das emitValue würde dann wirklich nur einfach den wert dispatchen und du kannst handleExternalFileSelection dann auch kicken.

Reply:
Siehe vorherigen Kommentar, bzw. müsste ich den store so umbauen, dass er mit mehreren `FileChooser` Instanzen gleichzeitig umgehen kann. Geht natürlich auch, braucht nur vermutlich jede Instanz eine id/key?

Der state hier erscheint mir gefühlt ausreichend lokal um ihn da zu lassen, aber du scheinst im Zweifel eher alles in den store zu packen, oder? :D Hast du Argumente für/gegen die Vorgehensweisen?

Reply:
Das würde die Komponente viel sauberer machen und den Daten flow in der saga transparenter. Lokalen state würde ich so gut es geht vermeiden. Gibt Sachen wie popover Sichtbarkeit da kann man dürber argumentieren :D

Reply:
Hab's umgebaut: fa52a0f8d3725d2c6896459482e0daa9fc897dd1

Naming könnte an einigen Stellen besser sein, da muss ich noch mal drüber. Schön ist auf jeden Fall, dass ich die `nextTick` workarounds nicht mehr brauche. Und ist nun natürlich leicht von überall nutzbar weil ich mich nicht drum kümmern muss, dass der lokale state zur entsprechenden Komponente kommt.

Gefühlt steckt allmählich zu viel in dem store/saga. Würdest du das aufsplitten, und wie? `auphonic.fileselection.store.ts`? `auphonic/fileselection.store.ts`?